src/org/freenetproject/freemail/wot/WoTConnections.java

```diff
@@ -20,10 +20,399 @@
+        returnWoTConnection(wotConnection);
+        return untrustedIdentities;
+    }
+
+    @Override
+    public Identity getIdentity(String identityId, String trusterId)
+            throws PluginNotFoundException, IOException, WoTException, InterruptedException {
+        WoTConnectionImpl wotConnection = borrowWoTConnection();
+        Identity wotIdentity = wotConnection.getIdentity(identityId, trusterId);
+        returnWoTConnection(wotConnection);
+        return wotIdentity;
+    }
+
+    @Override
+    public boolean setProperty(String identity, String key, String value)
+            throws PluginNotFoundException, IOException, InterruptedException {
+        WoTConnectionImpl wotConnection;
+        try {
+            wotConnection = borrowWoTConnection();
+        } catch (WoTException e) {
+            Logger.warning(this, "setProperty " + key + " = " + value + " failed", e);
+            return false;
+        }
+        boolean isSuccessfullySet = wotConnection.setProperty(identity, key, value);
+        returnWoTConnection(wotConnection);
+        return isSuccessfullySet;
+    }
+
+    @Override
+    public String getProperty(String identity, String key)
+            throws PluginNotFoundException, IOException, WoTException, InterruptedException {
+        WoTConnectionImpl wotConnection = borrowWoTConnection();
+        String parameterValue = wotConnection.getProperty(identity, key);
+        returnWoTConnection(wotConnection);
+        return parameterValue;
+    }
+
+    @Override
+    public boolean setContext(String identity, String context)
+            throws PluginNotFoundException, IOException, InterruptedException {
+        WoTConnectionImpl wotConnection;
+        try {
+            wotConnection = borrowWoTConnection();
+        } catch (WoTException e) {
+            Logger.warning(this, "setContext failed: " + identity + " - " + context, e);
+            return false;
+        }
+        boolean isSuccessfullySet = wotConnection.setContext(identity, context);
+        returnWoTConnection(wotConnection);
+        return isSuccessfullySet;
+    }
+
+    private WoTConnectionImpl borrowWoTConnection()
+            throws PluginNotFoundException, IOException, WoTException, InterruptedException {
+        WoTConnectionImpl wotConnection;
+        if ((wotConnection = wotConnectionsPool.poll()) == null) {
+            wotConnection = new WoTConnectionImpl(pluginRespirator);
+        } else {
+            try {
+                wotConnection.ping();
+            } catch (IOException | WoTException ignored) {
+                if (wotConnectionsPool.size() > 0) {
+                    borrowWoTConnection();
+                } else {
+                    wotConnection = new WoTConnectionImpl(pluginRespirator);
+                    wotConnection.ping();
+                }
+            }
+        }
+
+        return wotConnection;
+    }
```

Comment:
Also, this whole method seems way too over-complicated, especially that recursive call in there whose result is ignored. If we have a free connection object, use that, otherwise create a new one and use that…